lib/twilio-ruby/security/request_validator.rb

```diff
@@ -10,8 +10,20 @@ def initialize(auth_token = nil)
 
       def validate(url, params, signature)
         params_hash = params_to_hash(params)
-        expected = build_signature_for(url, params_hash)
-        secure_compare(expected, signature)
+        if params_hash.is_a? Enumerable
+          expected = build_signature_for(url, params_hash)
+          secure_compare(expected, signature)
+        else
+          expected_signature = build_signature_for(url, {})
+          body_hash = URI.decode_www_form(URI(url).query).to_h['bodySHA256']
+          expected_hash = build_hash_for(params)
+          secure_compare(expected_signature, signature) && secure_compare(expected_hash, body_hash)
+        end
+      end
+
+      def build_hash_for(body)
```

Comment:
I think we should disambiguate between `build_hash_for` and `params_to_hash` -- the latter could be `params_to_hashmap`; not sure about the former, maybe `build_token_for`?

Also I wonder if there is a better name for `params`, to indicate that it could also be a POST body. (I think `params` implies that we're talking about query parameters, but maybe I'm wrong about that.) Is `data` too lame?

Reply:
I wanted to rename it, but I think Ruby, like Python, lets you specify parameter names, so I was trying to do this backwards-compatibly. I guess we could rename it to `data` and make a separate `params` param at the end of the method to capture when it's called as `do_something(params={:some => 'stuff'})`. But I don't think you can set a default value for a parameter that comes before a non-default-able parameter, so you'd get missing parameter errors. In other words, this was my sort-of solution before in pseudo-code:

```
def validate(url, data, signature, params={})
  if data is null, use params
  else parse data if string or !string
end
```

But since it's _either_ data _or_ params, we'd need to change the parameter ordering to allow a default value for data. So at that point, it's backwards incompatible anyway.

I ran into the same issue with Python.

WDYT?

Reply:
ohhhhhhhkay yeah that is dumb. Too bad it's not at the end of the list.

Then I agree that it's best to keep the parameter name the same, but I also don't think we need to maintain backwards compatibility with the helper functions, so it should be OK to rename `params_to_hash`.

Probably there should be a docstring, though, right?